freqtrade/optimize/backtesting.py

```diff
@@ -283,6 +287,9 @@ def _get_ohlcv_as_lists(self, processed: Dict[str, DataFrame]) -> Dict[str, Tupl
             # Convert from Pandas to list for performance reasons
             # (Looping Pandas is slow.)
             data[pair] = df_analyzed[headers].values.tolist()
+
+            # Do not hold on to old data to reduce memory usage
+            processed[pair] = pair_data = DataFrame()
```

Comment:
would explicit delete statements (`del pair_data`, ...) provide even better performance?

i assume even creating an empty DataFrame has "some" overhead - especially if the purpose is to simply delete the origianl instance

Reply:
I tried that, it somehow made bunch of tests to fail with not quite obvious exceptions 👀 while in theory you are right, `DataFrame` object is bigger than no object, it is nothing compared to rest of the memory usage. Say we guestimate empty dataframe to be ~1KB (i doubt it is that big). 120 hyperopted pairs is a stretch, so overhead is 120KB which is nothing.

Reply:
how about assigning None?

that would not remove the entry from the dict - but also not create an unnecessary (and unusable!) dataframe

Reply:
Yeah i did not do that because typehint does not allow `None`. I thought type checking will bark. Lets see if that happens 👍🏻

Reply:
Hah looks like it worked 🥳

Reply:
Did you profile that latest iteration (reverting back to `items()`?)

The original profile showed quite some increase in that line (i suspect it's the aggregation of the content of all loops)- but i'd want to avoid "reverting" some benefits just because of this.

Reply:
<details>
  <summary>No revert:</summary>

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   245  11649.1 MiB  11649.1 MiB           1       @profile
   246                                             def _get_ohlcv_as_lists(self, processed: Dict[str, DataFrame]) -> Dict[str, Tuple]:
   247                                                 """
   248                                                 Helper function to convert a processed dataframes into lists for performance reasons.
   249                                         
   250                                                 Used by backtest() - so keep this optimized for performance.
   251                                         
   252                                                 :param processed: a processed dictionary with format {pair, data}, which gets cleared to
   253                                                 optimize memory usage!
   254                                                 """
   255                                                 # Every change to this headers list must evaluate further usages of the resulting tuple
   256                                                 # and eventually change the constants for indexes at the top
   257  11649.1 MiB      0.0 MiB           1           headers = ['date', 'buy', 'open', 'close', 'sell', 'low', 'high', 'buy_tag', 'exit_tag']
   258  11649.1 MiB      0.0 MiB           1           data: Dict = {}
   259  11649.1 MiB      0.0 MiB           1           self.progress.init_step(BacktestState.CONVERT, len(processed))
   260                                         
   261                                                 # Create dict with data
   262  15556.0 MiB   -739.2 MiB          46           for pair in processed.keys():
   263  15449.5 MiB   -739.2 MiB          45               pair_data = processed[pair]
   264  15449.5 MiB   -739.2 MiB          45               self.check_abort()
   265  15449.5 MiB   -739.2 MiB          45               self.progress.increment()
   266  15449.5 MiB   -739.2 MiB          45               if not pair_data.empty:
   267  15449.5 MiB   -739.2 MiB          45                   pair_data.loc[:, 'buy'] = 0  # cleanup if buy_signal is exist
   268  15449.5 MiB   -739.2 MiB          45                   pair_data.loc[:, 'sell'] = 0  # cleanup if sell_signal is exist
   269  15449.5 MiB   -739.2 MiB          45                   pair_data.loc[:, 'buy_tag'] = None  # cleanup if buy_tag is exist
   270  15449.5 MiB   -739.2 MiB          45                   pair_data.loc[:, 'exit_tag'] = None  # cleanup if exit_tag is exist
   271                                         
   272  15449.5 MiB  -1309.8 MiB          90               df_analyzed = self.strategy.advise_sell(
   273  15449.5 MiB   -739.2 MiB          45                   self.strategy.advise_buy(pair_data, {'pair': pair}), {'pair': pair}).copy()
   274                                                     # Trim startup period from analyzed dataframe
   275  15446.5 MiB  -1944.6 MiB          90               df_analyzed = trim_dataframe(df_analyzed, self.timerange,
   276  15446.5 MiB   -264.9 MiB          45                                            startup_candles=self.required_startup)
   277                                                     # To avoid using data from future, we use buy/sell signals shifted
   278                                                     # from the previous candle
   279  15446.4 MiB   -270.7 MiB          45               df_analyzed.loc[:, 'buy'] = df_analyzed.loc[:, 'buy'].shift(1)
   280  15446.4 MiB   -264.9 MiB          45               df_analyzed.loc[:, 'sell'] = df_analyzed.loc[:, 'sell'].shift(1)
   281  15446.4 MiB   -264.9 MiB          45               df_analyzed.loc[:, 'buy_tag'] = df_analyzed.loc[:, 'buy_tag'].shift(1)
   282  15446.4 MiB   -264.9 MiB          45               df_analyzed.loc[:, 'exit_tag'] = df_analyzed.loc[:, 'exit_tag'].shift(1)
   283                                         
   284                                                     # Update dataprovider cache
   285  15446.4 MiB   -264.9 MiB          45               self.dataprovider._set_cached_df(pair, self.timeframe, df_analyzed)
   286                                         
   287  15729.4 MiB  10179.0 MiB          45               df_analyzed = df_analyzed.drop(df_analyzed.head(1).index)
   288                                         
   289                                                     # Convert from Pandas to list for performance reasons
   290                                                     # (Looping Pandas is slow.)
   291  15832.7 MiB   1876.4 MiB          45               data[pair] = df_analyzed[headers].values.tolist()
   292                                         
   293                                                     # Do not hold on to old data to reduce memory usage
   294  15556.0 MiB -12697.3 MiB          45               processed[pair] = pair_data = None
   295  15556.0 MiB      0.0 MiB           1           return data
```

</details>

<details>
  <summary>With revert:</summary>

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
   245  11566.8 MiB  11566.8 MiB           1       @profile
   246                                             def _get_ohlcv_as_lists(self, processed: Dict[str, DataFrame]) -> Dict[str, Tuple]:
   247                                                 """
   248                                                 Helper function to convert a processed dataframes into lists for performance reasons.
   249                                         
   250                                                 Used by backtest() - so keep this optimized for performance.
   251                                         
   252                                                 :param processed: a processed dictionary with format {pair, data}, which gets cleared to
   253                                                 optimize memory usage!
   254                                                 """
   255                                                 # Every change to this headers list must evaluate further usages of the resulting tuple
   256                                                 # and eventually change the constants for indexes at the top
   257  11566.8 MiB      0.0 MiB           1           headers = ['date', 'buy', 'open', 'close', 'sell', 'low', 'high', 'buy_tag', 'exit_tag']
   258  11566.8 MiB      0.0 MiB           1           data: Dict = {}
   259  11566.8 MiB      0.0 MiB           1           self.progress.init_step(BacktestState.CONVERT, len(processed))
   260                                         
   261                                                 # Create dict with data
   262  15544.8 MiB -12839.3 MiB          46           for pair, pair_data in processed.items():
   263  15436.8 MiB   -968.5 MiB          45               self.check_abort()
   264  15436.8 MiB   -968.5 MiB          45               self.progress.increment()
   265  15436.8 MiB   -968.5 MiB          45               if not pair_data.empty:
   266  15436.8 MiB   -968.5 MiB          45                   pair_data.loc[:, 'buy'] = 0  # cleanup if buy_signal is exist
   267  15436.8 MiB   -968.5 MiB          45                   pair_data.loc[:, 'sell'] = 0  # cleanup if sell_signal is exist
   268  15436.8 MiB   -968.5 MiB          45                   pair_data.loc[:, 'buy_tag'] = None  # cleanup if buy_tag is exist
   269  15436.8 MiB   -968.5 MiB          45                   pair_data.loc[:, 'exit_tag'] = None  # cleanup if exit_tag is exist
   270                                         
   271  15436.8 MiB  -1765.9 MiB          90               df_analyzed = self.strategy.advise_sell(
   272  15436.8 MiB   -968.5 MiB          45                   self.strategy.advise_buy(pair_data, {'pair': pair}), {'pair': pair}).copy()
   273                                                     # Trim startup period from analyzed dataframe
   274  15433.7 MiB  -2180.9 MiB          90               df_analyzed = trim_dataframe(df_analyzed, self.timerange,
   275  15433.7 MiB   -501.2 MiB          45                                            startup_candles=self.required_startup)
   276                                                     # To avoid using data from future, we use buy/sell signals shifted
   277                                                     # from the previous candle
   278  15433.6 MiB   -280.1 MiB          45               df_analyzed.loc[:, 'buy'] = df_analyzed.loc[:, 'buy'].shift(1)
   279  15433.6 MiB   -274.4 MiB          45               df_analyzed.loc[:, 'sell'] = df_analyzed.loc[:, 'sell'].shift(1)
   280  15433.6 MiB   -274.4 MiB          45               df_analyzed.loc[:, 'buy_tag'] = df_analyzed.loc[:, 'buy_tag'].shift(1)
   281  15433.6 MiB   -274.4 MiB          45               df_analyzed.loc[:, 'exit_tag'] = df_analyzed.loc[:, 'exit_tag'].shift(1)
   282                                         
   283                                                     # Update dataprovider cache
   284  15433.6 MiB   -274.4 MiB          45               self.dataprovider._set_cached_df(pair, self.timeframe, df_analyzed)
   285                                         
   286  15716.6 MiB   9940.6 MiB          45               df_analyzed = df_analyzed.drop(df_analyzed.head(1).index)
   287                                         
   288                                                     # Convert from Pandas to list for performance reasons
   289                                                     # (Looping Pandas is slow.)
   290  15821.5 MiB   1615.0 MiB          45               data[pair] = df_analyzed[headers].values.tolist()
   291                                         
   292                                                     # Do not hold on to old data to reduce memory usage
   293  15821.5 MiB  -2710.6 MiB          45               processed[pair] = pair_data = None
   294  15544.8 MiB      0.0 MiB           1           return data
```

</details>

Seems like revert just moves up deallocation to the next loop iteration.

Reply:
that's a diff. test case now? number seem quite different to the original comment - and i'm too lazy to calculate relative difference :laughing:

Reply:
Its `for pair in processed.keys()` vs `for pair, pair_data in processed.items()`.

In `for pair in processed.keys()` variant bulk of deallocation is done on line 294:
```
   262  15556.0 MiB   -739.2 MiB          46           for pair in processed.keys():
   ...
   294  15556.0 MiB -12697.3 MiB          45               processed[pair] = pair_data = None
   295  15556.0 MiB      0.0 MiB           1           return data
```

In original `for pair, pair_data in processed.items()` bulk of deallocation is done when pair_data goes out of scope on line 262:
```
   262  15544.8 MiB -12839.3 MiB          46           for pair, pair_data in processed.items():
   ...
   293  15821.5 MiB  -2710.6 MiB          45               processed[pair] = pair_data = None
   294  15544.8 MiB      0.0 MiB           1           return data
```

End result is pretty much same. Now that i think of it there is no need to set `pair_data` to `None` after reverting loop change.

Reply:
> Now that i think of it there is no need to set pair_data to None after reverting loop change.

are we ok as is, or do you intend to remove that part? (I'd remove it if it's not necessary - but we should be sure it's not).

Reply:
I am profiling with it removed right now, hold your merge-happy finger for a moment :]

Reply:
Alright im done. Turns out `for pair in processed.keys()` loop change does help a little bit, it saves 4GB or so of my RAM usage in test i ran. Not a lot compared to 90GB used RAM, but still something.

Reply:
That was my question / doubt all along ;)